File: sdk/typescript/src/namespaces/relationships.ts
/**
 * Relationships Namespace API
 *
 * Provides access to agent-to-agent relationship data, including
 * collaboration patterns, rivalry analysis, and network graphs.
 */

import type { AragoraClient } from '../client';

/**
 * Individual agent relationship
 */
export interface AgentRelationship {
  agent_a: string;
  agent_b: string;
  relationship_type: 'ally' | 'rival' | 'neutral' | 'unknown';
  agreement_rate: number;
  total_interactions: number;
  last_interaction?: string;
  collaboration_score: number;
  conflict_score: number;
}

/**
 * Relationship graph node
 */
export interface RelationshipNode {
  agent: string;
  connections: number;
  centrality: number;
  cluster?: string;
}

/**
 * Relationship graph edge
 */
export interface RelationshipEdge {
  source: string;
  target: string;
  weight: number;
  type: 'positive' | 'negative' | 'neutral';
}

/**
 * Full relationship graph
 */
export interface RelationshipGraph {
  nodes: RelationshipNode[];
  edges: RelationshipEdge[];
  clusters: Array<{
    id: string;
    agents: string[];
    cohesion: number;
  }>;
  metadata: {
    total_agents: number;
    total_relationships: number;
    average_connections: number;
    generated_at: string;
  };
}

/**
 * Relationship statistics
 */
export interface RelationshipStats {
  total_relationships: number;
  ally_count: number;
  rival_count: number;
  neutral_count: number;
  average_agreement_rate: number;
  most_connected_agent: string;
  strongest_alliance: { agents: [string, string]; score: number };
  strongest_rivalry: { agents: [string, string]; score: number };
}

/**
 * Relationship summary
 */
export interface RelationshipSummary {
  agent: string;
  allies: string[];
  rivals: string[];
  neutral: string[];
  top_collaborators: Array<{ agent: string; score: number }>;
  summary_text: string;
}

/**
 * Relationships namespace for agent relationship analysis.
 *
 * @example
 * ```typescript
 * // Get relationship between two agents
 * const relationship = await client.relationships.get('gpt-4', 'claude');
 * console.log(`Agreement rate: ${relationship.agreement_rate}%`);
 *
 * // Get full relationship graph
 * const graph = await client.relationships.getGraph();
 * console.log(`Network has ${graph.nodes.length} agents`);
 * ```
 */
export class RelationshipsNamespace {
  constructor(private client: AragoraClient) {}

  /**
   * Get the relationship between two specific agents.
   *
   * @param agentA - First agent name
   * @param agentB - Second agent name
   * @returns Relationship details between the two agents
   */
  async get(agentA: string, agentB: string): Promise<AgentRelationship> {
    return this.client.request<AgentRelationship>(
      'GET',
      `/api/v1/relationship/${encodeURIComponent(agentA)}/${encodeURIComponent(agentB)}`
    );
  }

  /**
   * Get the full agent relationship graph.
   *
   * Useful for visualizing agent networks and identifying clusters.
   *
   * @param options.min_interactions - Minimum interactions to include edge
   * @param options.include_inactive - Include agents with no recent activity
   */
  async getGraph(options?: {
    min_interactions?: number;
    include_inactive?: boolean;
  }): Promise<RelationshipGraph> {
    return this.client.request<RelationshipGraph>('GET', '/api/v1/relationships/graph', {
      params: options,
    });
  }

  /**
   * Get aggregate relationship statistics.
   */
  async getStats(): Promise<RelationshipStats> {
    return this.client.request<RelationshipStats>('GET', '/api/v1/relationships/stats');
  }

  /**
   * Get a summary of relationships for a specific agent or all agents.
   *
   * @param agent - Optional agent name for agent-specific summary
   */
  async getSummary(agent?: string): Promise<RelationshipSummary | RelationshipSummary[]> {
    const params = agent ? { agent } : undefined;
    return this.client.request<RelationshipSummary | RelationshipSummary[]>(
      'GET',
      '/api/v1/relationships/summary',
      { params }
    );
  }
}

File: sdk/python/aragora/namespaces/relationships.py
"""
Relationships Namespace API

Provides access to agent relationship analysis, including rivalries and alliances.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class RelationshipsAPI:
    """Synchronous Relationships API for agent relationship analysis."""

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_summary(self) -> dict[str, Any]:
        """Get global relationship overview.

        Returns:
            Summary including strongest rivalry/alliance, most connected agent,
            and average scores.
        """
        return self._client.request("GET", "/api/v1/relationships/summary")

    def get_graph(
        self,
        min_debates: int = 3,
        min_score: float = 0.0,
    ) -> dict[str, Any]:
        """Get full relationship graph for visualizations.

        Args:
            min_debates: Minimum debates between agents to include (default: 3).
            min_score: Minimum rivalry/alliance score to include (default: 0.0).

        Returns:
            Graph structure with nodes (agents) and edges (relationships).
        """
        params = {"min_debates": min_debates, "min_score": min_score}
        return self._client.request("GET", "/api/v1/relationships/graph", params=params)

    def get_stats(self) -> dict[str, Any]:
        """Get relationship system statistics.

        Returns:
            Statistics including rivalry/alliance counts, most debated pair,
            and highest agreement pair.
        """
        return self._client.request("GET", "/api/v1/relationships/stats")

    def get_pair(
        self,
        agent_a: str,
        agent_b: str,
    ) -> dict[str, Any]:
        """Get detailed relationship between two specific agents.

        Args:
            agent_a: First agent name.
            agent_b: Second agent name.

        Returns:
            Detailed relationship including debate count, agreement rate,
            rivalry/alliance scores, head-to-head record, and influence metrics.
        """
        return self._client.request("GET", f"/api/v1/relationship/{agent_a}/{agent_b}")


class AsyncRelationshipsAPI:
    """Asynchronous Relationships API for agent relationship analysis."""

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_summary(self) -> dict[str, Any]:
        """Get global relationship overview."""
        return await self._client.request("GET", "/api/v1/relationships/summary")

    async def get_graph(
        self,
        min_debates: int = 3,
        min_score: float = 0.0,
    ) -> dict[str, Any]:
        """Get full relationship graph for visualizations."""
        params = {"min_debates": min_debates, "min_score": min_score}
        return await self._client.request("GET", "/api/v1/relationships/graph", params=params)

    async def get_stats(self) -> dict[str, Any]:
        """Get relationship system statistics."""
        return await self._client.request("GET", "/api/v1/relationships/stats")

    async def get_pair(
        self,
        agent_a: str,
        agent_b: str,
    ) -> dict[str, Any]:
        """Get detailed relationship between two specific agents."""
        return await self._client.request("GET", f"/api/v1/relationship/{agent_a}/{agent_b}")
